contracts/data-storage/andromeda-counter/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, ensure, Storage, Addr};
+
+use andromeda_data_storage::counter::{ExecuteMsg, InstantiateMsg, QueryMsg, CounterRestriction};
+use andromeda_data_storage::counter::{
+    GetInitialAmountResponse, GetCurrentAmountResponse, GetIncreaseAmountResponse, GetDecreaseAmountResponse, GetRestrictionResponse,
+};
+use andromeda_std::{
+    ado_base::{InstantiateMsg as BaseInstantiateMsg, MigrateMsg},
+    ado_contract::ADOContract,
+    common::{context::ExecuteContext, encode_binary},
+    error::ContractError,
+};
+use cw_utils::nonpayable;
+
+use crate::state::{RESTRICTION, CURRENT_AMOUNT, DEFAULT_INITIAL_AMOUNT, DEFAULT_INCREASE_AMOUNT, DEFAULT_DECREASE_AMOUNT, INITIAL_AMOUNT, INCREASE_AMOUNT, DECREASE_AMOUNT};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:andromeda-counter";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let resp = ADOContract::default().instantiate(
+        deps.storage,
+        env,
+        deps.api,
+        &deps.querier,
+        info,
+        BaseInstantiateMsg {
+            ado_type: CONTRACT_NAME.to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            kernel_address: msg.kernel_address,
+            owner: msg.owner,
+        },
+    )?;
+
+    RESTRICTION.save(deps.storage, &msg.restriction)?;
+
+    if let Some(initial_amount) = msg.initial_amount {
+        INITIAL_AMOUNT.save(deps.storage, &initial_amount)?;
+        CURRENT_AMOUNT.save(deps.storage, &initial_amount)?;
+    } else {
+        INITIAL_AMOUNT.save(deps.storage, &DEFAULT_INITIAL_AMOUNT)?;
+        CURRENT_AMOUNT.save(deps.storage, &DEFAULT_INITIAL_AMOUNT)?;
+    }
+
+    if let Some(increase_amount) = msg.increase_amount {
+        INCREASE_AMOUNT.save(deps.storage, &increase_amount)?;
+    } else {
+        INCREASE_AMOUNT.save(deps.storage, &DEFAULT_INCREASE_AMOUNT)?;
+    }
+
+    if let Some(decrease_amount) = msg.decrease_amount {
+        DECREASE_AMOUNT.save(deps.storage, &decrease_amount)?;
+    } else {
+        DECREASE_AMOUNT.save(deps.storage, &DEFAULT_DECREASE_AMOUNT)?;
+    }
+
+    Ok(resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let ctx = ExecuteContext::new(deps, info, env);
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        },
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+fn handle_execute(ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+
```

Comment:
You can add this here to avoid checking for permissions for every function separately
```rust
    let action = get_action_name(CONTRACT_NAME, msg.as_ref());

    let action_response = call_action(
        &mut ctx.deps,
        &ctx.info,
        &ctx.env,
        &ctx.amp_ctx,
        msg.as_ref(),
    )?;
```